src/CheckConcreteStructs.jl

```diff
@@ -129,7 +129,14 @@ function recursive_isconcretetype(T)
     isa(T, TypeVar) && return true  # key modification
     isconcretetype(T) && return true
     isabstracttype(T) && return false
-    for subT in fieldtypes(T)
+    if isa(T, Union)
+        subTs = Base.uniontypes(T)
+        # If inference will give up on this union, report as `false`.
+        length(subTs) > Core.Compiler.MAX_TYPEUNION_LENGTH && return false
```

Comment:
I'm not sure I agree with this. First of all, it's an internal so it could break (e.g. be renamed) in a future release. Second, we're not checking inferrability here, we're checking concreteness. And it seems non-trivial `Union` types are never concrete:
```julia
julia> isconcretetype(Union{Int})
true

julia> isconcretetype(Union{Int,Float64})
false
```
So maybe we should just return `false` if the length of the `Union` is > 1? Or have some additional "sensitivity / log level" macro parameter which we can use to ignore harmless stuff like small unions. What do you think?

Reply:
I'm gonna merge because it fixes a bug but I'm still not completely convinced that this is the right heuristic

Reply:
Yeah, I wasn't completely satisfied with this for the reasons you suggest.

It seems to me however that unions are completely reasonable to have in good quality code, as they're quite decent performance-wise. So although they're not concrete, I was thinking they probably shouldn't be forbidden by default unless someone takes a large union that would undoubtedly have significant performance drawbacks.

At the very least, I think unions with 1 or 2 parameters are fine, and from 3 it would make sense that we start returning `false` (even if it's only starting from 4 that performance should really be impacted considerably).

We can hardcode ` length(subTs) > n` with e.g. `n = 2` or `n = 3` (as the compiler heuristic for inferability will not likely change, and if it does, at least we're not accessing an internal symbol).

If you still prefer to go with not supporting non-concrete `Union`s by default, which I would completely understand given the purpose of the package (checking for concreteness), we could support them with some opt-in mechanism ("I know that using `Union`s for my code is a sensible choice"), which could look like
```julia
@check_concrete struct MyType
  @skipcheck x::Union{A, B, C, D, E, F}
end
```

Those are just a few thoughts, I'm good with any of the suggestions that have been made so far.

Reply:
Let's keep it that way for now, it's very simple and I think it does "the right thing" as far as beginners are concerned (telling them that a union of 2 things is ok but a union of 10 isn't).
I have opened #3 to keep track of this discussion, and in case it comes up again.